File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.java
import java.util.*;
import java.io.*;

class Main {
	
	public static void main(String[] args) throws IOException {
		
		// N, M 초기화
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		
		// 수열 초기화
		int[] array = new int[N];
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			array[i] = Integer.parseInt(st.nextToken());  
		}
		
		// 투포인터로 M이 되는 경우 구하기
		int left = 0;
		int right = 0;
		int count = 0;
		int sum = array[right];
		
		
		while (right < N) {
			
			if (sum == M) {
				count++;
			}
			
			
			if (sum <= M) {
				right++;
				if (right != N) {
					sum += array[right];
				}
			}
			
			else if (sum > M) {
				sum -= array[left];
				left++;
				
			}
			

		}
		
		System.out.println(count);
	}
	
}

File: 백준/Silver/2003. 수들의 합 2/수들의 합 2.cc
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

int main() {

    // 입력
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    vector<int> array(n);
    for (int i = 0; i < n; i++) {
        cin >> array[i];
    }

    // 투 포인터로 합이 m이 되는 경우 구하기
    int left = 0;
    int right = 0;
    int sum = array[left];
    int answer = sum == m ? 1: 0;
    while (right < n) {

        if (sum <= m) {
            if (right++ == n - 1) break;
            sum += array[right];
        }
        else if (sum > m) {
            sum -= array[left++];
        }

        if (sum == m) {
            answer++;
        }

    }


    // 출력
    cout << answer;
}